src/runner/browser-set.js

```diff
@@ -0,0 +1,153 @@
+import { EventEmitter } from 'events';
+import { Promise } from 'es6-promise';
+import once from 'once';
+import LocalBrowserConnection from '../browser-connection/local';
+import { MESSAGE, getText } from '../messages';
+import remove from '../utils/array-remove';
+
+
+export default class BrowserSet extends EventEmitter {
+    constructor (connections) {
+        super();
+
+        this.BROWSER_CONNECTION_READY_TIMEOUT = 30000;
+        this.WAITING_FOR_IDLE_TIMEOUT         = 10000;
+
+        this.promisedDisposals = [];
+
+        this.connections = connections;
+
+        this.browserErrorHandler = msg => this.emit('error', msg);
+
+        connections.forEach(bc => bc.on('error', this.browserErrorHandler));
+
+        //NOTE: We're setting an empty error handler, because Node kills the process on an 'error' event
+        //if there is no handler. See https://nodejs.org/api/events.html#events_class_events_eventemitter
+        this.on('error', () => {});
+    }
+
+    _waitIdle (bc) {
+        return new Promise(resolve => {
+            if (bc.idle || bc.closed || !bc.ready) {
+                resolve();
+                return;
+            }
+            var waitingForIdleTimeout = setTimeout(() => resolve(), this.WAITING_FOR_IDLE_TIMEOUT);
+
+            var eventHandler = once(() => {
+                clearTimeout(waitingForIdleTimeout);
+                resolve();
+            });
+
+            bc.once('idle', eventHandler);
+            bc.once('closed', eventHandler);
```

Comment:
If connection fails while we waiting for idle, it will be closed immediatelly by itself. So we need to attach a handler to close event.